File: Data Structures & Algorithms/palindrome-linked-list/submission-2.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    
    isPalindrome(head) {
        if (!head || !head.next) return true;
        
        // Step 1: Find the middle of the linked list
        let slow = head;
        let fast = head;
        while (fast !== null && fast.next !== null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // Step 2: Reverse the second half of the list
        let prev = null;
        let curr = slow;
        while (curr !== null) {
            let nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }
        
        // Step 3: Compare the first half and the reversed second half
        let firstHalf = head;
        let secondHalf = prev; // 'prev' is now the head of the reversed second half
        
        while (secondHalf !== null) {
            if (firstHalf.val !== secondHalf.val) {
                return false;
            }
            firstHalf = firstHalf.next;
            secondHalf = secondHalf.next;
        }
        
        return true;

}
}

File: Data Structures & Algorithms/add-two-numbers-ii/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        stack1 = []
        stack2 = []

        head = None
        carry = 0

        while l1 != None:
            stack1.append(l1.val)
            l1 = l1.next

        while l2 != None:
            stack2.append(l2.val)
            l2 = l2.next

        while len(stack1) > 0 or len(stack2) > 0 or carry > 0:
            sum = carry

            if len(stack1) > 0:
                sum += stack1.pop()

            if len(stack2) > 0:
                sum += stack2.pop()

            carry = sum // 10
            digit = sum % 10

            newNode = ListNode(digit)
            newNode.next = head
            head = newNode

        return head
        